processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -120,24 +120,20 @@ public void validateSortedEncodedArrays(
               dimValName
           );
         }
+      } else {
+        throw new SegmentValidationException(
+            "Dim [%s] value lengths not equal. Expected %d found %d",
+            dimensionName,
+            lhsLen,
+            rhsLen
+        );
       }
-    } else {
-      throw new SegmentValidationException(
-          "Dim [%s] value lengths not equal. Expected %d found %d",
-          dimensionName,
-          lhsLen,
-          rhsLen
-      );
     }
 
     for (int j = 0; j < Math.max(lhsLen, rhsLen); ++j) {
       final int dIdex1 = lhsLen <= j ? -1 : lhs[j];
       final int dIdex2 = rhsLen <= j ? -1 : rhs[j];
 
-      if (dIdex1 == dIdex2) {
-        continue;
-      }
-
       final String dim1ValName = dIdex1 < 0 ? null : lhsEncodings.get(dIdex1);
       final String dim2ValName = dIdex2 < 0 ? null : rhsEncodings.get(dIdex2);
       if ((dim1ValName == null) || (dim2ValName == null)) {
```

Comment:
Probably, but IMHO that's outside the scope of this PR, which is specifically about a bug fix.